File: JavaScript/max-continuous-series-of-1s.js
/* eslint-disable no-console */
// Brute force solution using nested loops
const maxConsectiveOne = (array, k) => {
  const n = array.length;
  let maxOneCount = 0;
  for (let i = 0; i < n; i + 1) {
    let zeroCount = 0;
    const j = i;
    for (j; j < n; j + 1) {
      if (array[j] === 0) {
        zeroCount += 1;
        if (zeroCount > k) break;
      }
    }
    maxOneCount = Math.max(maxOneCount, (j - (i + 1)) + 1);
  }
  return maxOneCount;
};

console.log(maxConsectiveOne([1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1], 2));
console.log(maxConsectiveOne([1, 1, 0, 1, 0, 1, 0, 0, 1], 1));

// Efficient solution using sliding window technique
const maxConsectiveOneUpdate = (array, k) => {
  const n = array.length;
  let leftEnd = 0;
  let maxOneCount = 0;
  let zeroCount = 0;
  for (let rightEnd = 0; rightEnd < n; rightEnd + 1) {
    if (array[rightEnd] === 0) {
      zeroCount += 1;
    }
    if (zeroCount > k) {
      if (array[leftEnd] === 0) {
        zeroCount -= 1;
      }
      leftEnd += 1;
    }
    maxOneCount = Math.max(maxOneCount, ((rightEnd - leftEnd) + 1));
  }
  return maxOneCount;
};

console.log(maxConsectiveOneUpdate([1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1], 2));
console.log(maxConsectiveOneUpdate([1, 1, 0, 1, 0, 1, 0, 0, 1], 1));

File: Ruby/algorithms/max_continuous_series_of_1s.rb
# Brute force solution using nested loops
def max_consective_one(array, zero_flips)
  n = array.length
  max_one_count = 0
  (0...n).each do |i|
    count_zeros = 0
    (i..n).each do |j|
      next unless (array[j]).zero?

      count_zeros += 1
      break if count_zeros > zero_flips
    end
    max_one_count = [max_one_count, (j - (i + 1)) + 1].max
  end
  max_one_count
end

# Efficient solution using sliding window technique
def max_consective_one_update(array, zero_flips)
  n = array.length
  max_one_count = 0
  zero_count = 0
  left_end = 0
  (0...n).each do |right_end|
    zero_count += 1 if (array[right_end]).zero?
    if zero_count > zero_flips
      zero_count -= 1 if (array[left_end]).zero?
      left_end += 1
    end
    max_one_count = [max_one_count, ((right_end - left_end) + 1)].max
  end
  max_one_count
end

p max_consective_one_update([1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1], 2)

p max_consective_one_update([1, 1, 0, 1, 0, 1, 0, 0, 1], 1)
